// eslint-config-plugin/index.js
module.exports = {
  root: true,
  env: {
    browser: true,
    node: true,
    es6: true,
    jest: true,
  },
  extends: [
    'airbnb-typescript',
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react-redux/recommended',
    'prettier',
    'prettier/@typescript-eslint',
    'prettier/babel',
    'prettier/react',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    project: ["node_modules/eslint-config-plugin/tsconfig.eslint.json"],
    sourceType: 'module',
    ecmaVersion: 2018,
    ecmaFeatures: {
      jsx: true,
    },
  },
  settings: {
    react: {
      version: 'detect',
    },
    'import/resolver': {
      node: {
        extensions: ['.ts', '.tsx', '.js', '.jsx'],
        paths: ['src'],
      },
    },
  },
  plugins: [
    '@typescript-eslint',
    '@typescript-eslint/eslint-plugin',
    'eslint-plugin-import',
    'eslint-plugin-react',
    'eslint-plugin-react-hooks',
    'react',
    'react-redux',
    'prettier',
  ],
  rules: {
    'import/order': 'off',
    'import/no-named-as-default': 'off',
    'no-use-before-define': 'off',
    'class-methods-use-this': 'off',
    'implicit-arrow-linebreak': 'off',
    'linebreak-style': 'off',
    'lines-between-class-members': 'off',
    'no-console': ['error', { allow: ['error', 'warning'] }],
    'no-empty': ['error', { allowEmptyCatch: true }],
    'no-plusplus': 'off',
    'no-restricted-syntax': 'off',
    'no-shadow': 'off',
    'no-underscore-dangle': 'off',
    'object-curly-newline': ['error', { multiline: true, consistent: true }],
    'operator-linebreak': ['error', 'after', { overrides: { '=': 'after', '?': 'before', ':': 'before' } }],
    'no-nested-ternary': 'error',
    'template-curly-spacing': ['error', 'never'],
    'comma-dangle': 'off',
    'consistent-return': 'off',

    '@typescript-eslint/no-use-before-define': ['error'],
    '@typescript-eslint/explicit-function-return-type': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/no-namespace': 'off',
    '@typescript-eslint/no-empty-function': 'off',
    '@typescript-eslint/no-empty-interface': 'off',
    '@typescript-eslint/no-shadow': 'off',
    '@typescript-eslint/no-unused-vars': 'error',
    '@typescript-eslint/no-inferrable-types': 'off',
    '@typescript-eslint/lines-between-class-members': 'off',

    'import/extensions': 'off',
    'import/no-extraneous-dependencies': 'off',
    'import/prefer-default-export': 'off',

    'jsx-a11y/anchor-is-valid': 'off',
    'jsx-a11y/click-events-have-key-events': 'off',
    'jsx-a11y/no-static-element-interactions': 'off',

    'react/destructuring-assignment': 'off',
    'react/jsx-curly-spacing': [2, 'never'],
    'react/no-array-index-key': 'off',
    'react/prop-types': 'off',
    'react/sort-comp': 'off',
    'react/static-property-placement': 'off',
    'react/jsx-props-no-spreading': 'off',
    'react/require-default-props': 'off',

    'react-redux/connect-prefer-named-arguments': 'off',
    'react-redux/prefer-separate-component-file': 'off',
    'react-redux/mapDispatchToProps-prefer-shorthand': 'off',

    'prettier/prettier': 'error',
  },
};
